File: tests/udp_control_test.js
#!/usr/bin/env node
/**
 * UDP Control Test Tool for live777 DataChannel Bridge (Node.js version)
 * 
 * Usage:
 *   node tests/udp_control_test.js --host 127.0.0.1 --port 5005 --message "test"
 */

const dgram = require('dgram');
const readline = require('readline');

class UdpControlTester {
  constructor(host = '127.0.0.1', port = 5005) {
    this.host = host;
    this.port = port;
    this.client = dgram.createSocket('udp4');
  }

  sendMessage(message, callback) {
    const buffer = Buffer.from(message);
    this.client.send(buffer, this.port, this.host, (err) => {
      if (err) {
        console.error(`✗ Error sending message: ${err.message}`);
      } else {
        console.log(`✓ Sent message: ${message}`);
      }
      if (callback) callback(err);
    });
  }

  sendJson(data, callback) {
    const message = JSON.stringify(data);
    this.sendMessage(message, callback);
  }

  sendBinary(hexString, callback) {
    const buffer = Buffer.from(hexString, 'hex');
    this.client.send(buffer, this.port, this.host, (err) => {
      if (err) {
        console.error(`✗ Error sending binary: ${err.message}`);
      } else {
        console.log(`✓ Sent binary: ${hexString}`);
      }
      if (callback) callback(err);
    });
  }

  sendPtzCommand(action, params = {}, callback) {
    const command = { action, ...params };
    this.sendJson(command, callback);
  }

  async stressTest(count, interval = 10) {
    console.log(`\n=== Stress Test ===`);
    console.log(`Sending ${count} messages to ${this.host}:${this.port}`);
    console.log(`Interval: ${interval}ms\n`);

    let success = 0;
    let failed = 0;

    for (let i = 0; i < count; i++) {
      const message = JSON.stringify({
        seq: i,
        timestamp: Date.now()
      });

      await new Promise((resolve) => {
        this.sendMessage(message, (err) => {
          if (err) {
            failed++;
          } else {
            success++;
          }
          
          if ((i + 1) % 100 === 0) {
            console.log(`Sent ${i + 1}/${count} messages...`);
          }

          setTimeout(resolve, interval);
        });
      });
    }

    console.log(`\n=== Results ===`);
    console.log(`Success: ${success}`);
    console.log(`Failed: ${failed}`);
    console.log(`Success rate: ${(success / count * 100).toFixed(2)}%`);
  }

  interactiveMode() {
    console.log(`\n=== UDP Control Interactive Mode ===`);
    console.log(`Target: ${this.host}:${this.port}`);
    console.log(`Commands:`);
    console.log(`  text <message>     - Send text message`);
    console.log(`  json <json_string> - Send JSON message`);
    console.log(`  pan <left|right>   - Pan camera`);
    console.log(`  tilt <up|down>     - Tilt camera`);
    console.log(`  zoom <in|out>      - Zoom camera`);
    console.log(`  quit               - Exit`);
    console.log();

    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
      prompt: 'udp> '
    });

    rl.prompt();

    rl.on('line', (line) => {
      const input = line.trim();
      if (!input) {
        rl.prompt();
        return;
      }

      const parts = input.split(/\s+/);
      const command = parts[0].toLowerCase();
      const args = parts.slice(1);

      switch (command) {
        case 'quit':
        case 'exit':
          console.log('Exiting...');
          rl.close();
          this.client.close();
          return;

        case 'text':
          if (args.length > 0) {
            this.sendMessage(args.join(' '));
          } else {
            console.log('✗ Usage: text <message>');
          }
          break;

        case 'json':
          if (args.length > 0) {
            try {
              const data = JSON.parse(args.join(' '));
              this.sendJson(data);
            } catch (e) {
              console.log(`✗ Invalid JSON: ${e.message}`);
            }
          } else {
            console.log('✗ Usage: json <json_string>');
          }
          break;

        case 'pan':
          if (args.length > 0 && ['left', 'right'].includes(args[0])) {
            this.sendPtzCommand('pan', { direction: args[0], speed: 50 });
          } else {
            console.log('✗ Usage: pan <left|right>');
          }
          break;

        case 'tilt':
          if (args.length > 0 && ['up', 'down'].includes(args[0])) {
            this.sendPtzCommand('tilt', { direction: args[0], speed: 50 });
          } else {
            console.log('✗ Usage: tilt <up|down>');
          }
          break;

        case 'zoom':
          if (args.length > 0 && ['in', 'out'].includes(args[0])) {
            this.sendPtzCommand('zoom', { direction: args[0], value: 1 });
          } else {
            console.log('✗ Usage: zoom <in|out>');
          }
          break;

        default:
          console.log(`✗ Unknown command: ${command}`);
      }

      rl.prompt();
    });

    rl.on('close', () => {
      this.client.close();
      process.exit(0);
    });
  }

  close() {
    this.client.close();
  }
}

// Parse command line arguments
function parseArgs() {
  const args = process.argv.slice(2);
  const options = {
    host: '127.0.0.1',
    port: 5005,
    message: null,
    json: null,
    binary: null,
    interactive: false,
    stress: null,
    interval: 10
  };

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '--host':
        options.host = args[++i];
        break;
      case '--port':
        options.port = parseInt(args[++i]);
        break;
      case '--message':
        options.message = args[++i];
        break;
      case '--json':
        options.json = args[++i];
        break;
      case '--binary':
        options.binary = args[++i];
        break;
      case '--interactive':
        options.interactive = true;
        break;
      case '--stress':
        options.stress = parseInt(args[++i]);
        break;
      case '--interval':
        options.interval = parseFloat(args[++i]);
        break;
      case '--help':
        console.log('UDP Control Test Tool');
        console.log('\nOptions:');
        console.log('  --host <host>       Target host (default: 127.0.0.1)');
        console.log('  --port <port>       Target port (default: 5005)');
        console.log('  --message <text>    Send a text message');
        console.log('  --json <json>       Send a JSON message');
        console.log('  --binary <hex>      Send binary data (hex string)');
        console.log('  --interactive       Enter interactive mode');
        console.log('  --stress <count>    Stress test with COUNT messages');
        console.log('  --interval <ms>     Interval between messages (default: 10ms)');
        console.log('  --help              Show this help');
        process.exit(0);
    }
  }

  return options;
}

async function main() {
  const options = parseArgs();
  const tester = new UdpControlTester(options.host, options.port);

  if (options.message) {
    tester.sendMessage(options.message, () => tester.close());
  } else if (options.json) {
    try {
      const data = JSON.parse(options.json);
      tester.sendJson(data, () => tester.close());
    } catch (e) {
      console.error(`✗ Invalid JSON: ${e.message}`);
      tester.close();
      process.exit(1);
    }
  } else if (options.binary) {
    tester.sendBinary(options.binary, () => tester.close());
  } else if (options.stress) {
    await tester.stressTest(options.stress, options.interval);
    tester.close();
  } else if (options.interactive) {
    tester.interactiveMode();
  } else {
    console.log('UDP Control Test Tool');
    console.log('\nExamples:');
    console.log(`  node ${process.argv[1]} --message "Hello"`);
    console.log(`  node ${process.argv[1]} --json '{"action":"pan","direction":"left"}'`);
    console.log(`  node ${process.argv[1]} --binary "010032"`);
    console.log(`  node ${process.argv[1]} --interactive`);
    console.log(`  node ${process.argv[1]} --stress 1000 --interval 10`);
    console.log('\nUse --help for more options');
    tester.close();
  }
}

main().catch((err) => {
  console.error('Error:', err);
  process.exit(1);
});

File: tests/udp_control_test.py
#!/usr/bin/env python3
"""
UDP Control Test Tool for live777 DataChannel Bridge

Usage:
    python tests/udp_control_test.py --host 127.0.0.1 --port 5005 --message "test"
"""

import socket
import argparse
import json
import time
import sys


def send_text_message(host: str, port: int, message: str):
    """Send a text message via UDP"""
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        sock.sendto(message.encode('utf-8'), (host, port))
        print(f"✓ Sent text message: {message}")
    except Exception as e:
        print(f"✗ Error sending message: {e}", file=sys.stderr)
    finally:
        sock.close()


def send_json_message(host: str, port: int, data: dict):
    """Send a JSON message via UDP"""
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        message = json.dumps(data).encode('utf-8')
        sock.sendto(message, (host, port))
        print(f"✓ Sent JSON message: {json.dumps(data, indent=2)}")
    except Exception as e:
        print(f"✗ Error sending JSON: {e}", file=sys.stderr)
    finally:
        sock.close()


def send_binary_message(host: str, port: int, data: bytes):
    """Send binary data via UDP"""
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        sock.sendto(data, (host, port))
        print(f"✓ Sent binary message: {data.hex()}")
    except Exception as e:
        print(f"✗ Error sending binary: {e}", file=sys.stderr)
    finally:
        sock.close()


def send_ptz_command(host: str, port: int, action: str, **kwargs):
    """Send PTZ control command"""
    command = {"action": action, **kwargs}
    send_json_message(host, port, command)


def interactive_mode(host: str, port: int):
    """Interactive mode for sending commands"""
    print(f"\n=== UDP Control Interactive Mode ===")
    print(f"Target: {host}:{port}")
    print(f"Commands:")
    print(f"  text <message>     - Send text message")
    print(f"  json <json_string> - Send JSON message")
    print(f"  pan <left|right>   - Pan camera")
    print(f"  tilt <up|down>     - Tilt camera")
    print(f"  zoom <in|out>      - Zoom camera")
    print(f"  quit               - Exit")
    print()

    while True:
        try:
            cmd = input("udp> ").strip()
            if not cmd:
                continue

            parts = cmd.split(maxsplit=1)
            command = parts[0].lower()

            if command == "quit":
                break
            elif command == "text" and len(parts) > 1:
                send_text_message(host, port, parts[1])
            elif command == "json" and len(parts) > 1:
                try:
                    data = json.loads(parts[1])
                    send_json_message(host, port, data)
                except json.JSONDecodeError as e:
                    print(f"✗ Invalid JSON: {e}")
            elif command == "pan" and len(parts) > 1:
                direction = parts[1].lower()
                if direction in ["left", "right"]:
                    send_ptz_command(host, port, "pan", direction=direction, speed=50)
                else:
                    print("✗ Invalid direction. Use: left or right")
            elif command == "tilt" and len(parts) > 1:
                direction = parts[1].lower()
                if direction in ["up", "down"]:
                    send_ptz_command(host, port, "tilt", direction=direction, speed=50)
                else:
                    print("✗ Invalid direction. Use: up or down")
            elif command == "zoom" and len(parts) > 1:
                direction = parts[1].lower()
                if direction in ["in", "out"]:
                    send_ptz_command(host, port, "zoom", direction=direction, value=1)
                else:
                    print("✗ Invalid direction. Use: in or out")
            else:
                print(f"✗ Unknown command: {command}")

        except KeyboardInterrupt:
            print("\n\nExiting...")
            break
        except Exception as e:
            print(f"✗ Error: {e}")


def stress_test(host: str, port: int, count: int, interval: float):
    """Send multiple messages for stress testing"""
    print(f"\n=== Stress Test ===")
    print(f"Sending {count} messages to {host}:{port}")
    print(f"Interval: {interval}s\n")

    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    success = 0
    failed = 0

    try:
        for i in range(count):
            message = json.dumps({"seq": i, "timestamp": time.time()})
            try:
                sock.sendto(message.encode('utf-8'), (host, port))
                success += 1
                if (i + 1) % 100 == 0:
                    print(f"Sent {i + 1}/{count} messages...")
            except Exception as e:
                failed += 1
                print(f"✗ Failed to send message {i}: {e}")

            if interval > 0:
                time.sleep(interval)

    finally:
        sock.close()

    print(f"\n=== Results ===")
    print(f"Success: {success}")
    print(f"Failed: {failed}")
    print(f"Success rate: {success / count * 100:.2f}%")


def main():
    parser = argparse.ArgumentParser(
        description="UDP Control Test Tool for live777 DataChannel Bridge"
    )
    parser.add_argument(
        "--host",
        default="127.0.0.1",
        help="Target host (default: 127.0.0.1)"
    )
    parser.add_argument(
        "--port",
        type=int,
        default=5005,
        help="Target port (default: 5005)"
    )
    parser.add_argument(
        "--message",
        help="Send a single text message and exit"
    )
    parser.add_argument(
        "--json",
        help="Send a single JSON message and exit"
    )
    parser.add_argument(
        "--binary",
        help="Send binary data (hex string) and exit"
    )
    parser.add_argument(
        "--interactive",
        action="store_true",
        help="Enter interactive mode"
    )
    parser.add_argument(
        "--stress",
        type=int,
        metavar="COUNT",
        help="Stress test: send COUNT messages"
    )
    parser.add_argument(
        "--interval",
        type=float,
        default=0.01,
        help="Interval between messages in stress test (default: 0.01s)"
    )

    args = parser.parse_args()

    # Single message modes
    if args.message:
        send_text_message(args.host, args.port, args.message)
    elif args.json:
        try:
            data = json.loads(args.json)
            send_json_message(args.host, args.port, data)
        except json.JSONDecodeError as e:
            print(f"✗ Invalid JSON: {e}", file=sys.stderr)
            sys.exit(1)
    elif args.binary:
        try:
            data = bytes.fromhex(args.binary)
            send_binary_message(args.host, args.port, data)
        except ValueError as e:
            print(f"✗ Invalid hex string: {e}", file=sys.stderr)
            sys.exit(1)
    elif args.stress:
        stress_test(args.host, args.port, args.stress, args.interval)
    elif args.interactive:
        interactive_mode(args.host, args.port)
    else:
        # Default: show examples
        print("UDP Control Test Tool")
        print("\nExamples:")
        print(f"  {sys.argv[0]} --message 'Hello'")
        print(f"  {sys.argv[0]} --json '{{\"action\":\"pan\",\"direction\":\"left\"}}'")
        print(f"  {sys.argv[0]} --binary '010032'")
        print(f"  {sys.argv[0]} --interactive")
        print(f"  {sys.argv[0]} --stress 1000 --interval 0.01")
        print("\nUse --help for more options")


if __name__ == "__main__":
    main()
